### PR TITLE
Implement a holepunch transport

### LINKED ISSUE #264: Implement a holepunch transport
> We need a transport that uses a point to point connection between two parties that are behind firewalls.
> 
> There are standard protocols that we should be able to use. The main one is called ICE.

### PR DESCRIPTION
**Pull request by nkryuchkov**
Did you run `make format && make check`? 

Fixes #264

 Changes:	
- Implement `stcp` with address resolving (`stcpr`) and encryption
- Implement `stcp` with TCP hole punching (`stcph`) and encryption
- Add encryption for `stcp`

How to test this PR:
- Run generic env
- Run `make integration-startup-stcpr`
- Test if `skychat` works
- Shutdown env
- Do the same for `stcph` by running `make integration-startup-stcph`
- Change `address-resolver` addresses to `http://address-resolver.skywire.cc` in `Visor{A,B,C}` configs (this does not work with all NATs)
- Do the same again

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -52,43 +53,53 @@ $ skywire-cli visor gen-config
...
 ```json
 {
   "stcp": {
-    "pk_table": {
-      "024a2dd77de324d543561a6d9e62791723be26ddf6b9587060a10b9ba498e096f1": "127.0.0.1:7031",
-      "0327396b1241a650163d5bc72a7970f6dfbcca3f3d67ab3b15be9fa5c8da532c08": "127.0.0.1:7032"
-    },
-    "local_address": "127.0.0.1:7033"
+    "address_resolver": "http://localhost:9093",
+    "local_address": "localhost:7033"
```
Comment (Darkren) [inline README.md]: I guess we shouldn't do a bad example with this lol. Could you please remove localhost from `local_address` here?

File: internal/httpauth/client.go [hunk 2]
```diff
@@ -44,11 +45,13 @@ type HTTPError struct {
...
 type Client struct {
-	client http.Client
-	key    cipher.PubKey
-	sec    cipher.SecKey
-	addr   string // sanitized address of the client, which may differ from addr used in NewClient
-	nonce  uint64 // has to be handled with the atomic package at all time
+	nonce       uint64 // has to be handled with the atomic package at all time, needs to be 8 byte aligned
+	mu          sync.Mutex
+	client      *http.Client
+	reuseClient *http.Client
```
Comment (Darkren) [inline internal/httpauth/client.go:52]: I'm sorry but why do we need this one separately? not really clear

Comment (Darkren) [reply internal/httpauth/client.go:52]: I think I get it, but probably a good comment will do in here

File: internal/vpn/env.go [hunk 3]
```diff
@@ -25,6 +25,10 @@ const (
 	RFAddrEnvKey = "ADDR_RF"
 	// UptimeTrackerAddrEnvKey is env arg holding uptime tracker address.
 	UptimeTrackerAddrEnvKey = "ADDR_UPTIME_TRACKER"
+	// STCPRAddressResolverAddrEnvKey is env arg holding stcpr address resolver address.
+	STCPRAddressResolverAddrEnvKey = "STCPR_ADDR_ADDRESS_RESOLVER"
```
Comment (Darkren) [inline internal/vpn/env.go]: It should be `ADDR_STCPR_ADDRESS_RESOLVER`. same applies to `stcph`. it doesn't really matter, but the whole idea was to prefix the actual key with the `ADDR`. it'll just be in the same style as other keys above

File: pkg/snet/arclient/client.go [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+	}
+
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			log.WithError(err).Warn("Failed to close response body")
+		}
+	}()
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("status: %d, error: %v", resp.StatusCode, extractError(resp.Body))
```
Comment (Darkren) [inline pkg/snet/arclient/client.go]: why don't we use `%w` here?

Comment (Darkren) [reply pkg/snet/arclient/client.go]: And in the below funcs too

File: pkg/snet/arclient/client.go [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+	PK   cipher.PubKey
+	Addr string
+}
+
+func (c *httpClient) WS(ctx context.Context, dialCh <-chan cipher.PubKey) (<-chan RemoteVisor, error) {
+	addrCh := make(chan RemoteVisor, addrChSize)
+
+	if c.wsConn != nil {
+		if err := c.wsConn.Close(websocket.StatusNormalClosure, "new connection created"); err != nil {
+			_ = err // TODO: log error
```
Comment (Darkren) [inline pkg/snet/arclient/client.go]: should we log it?

File: pkg/snet/network.go [hunk 6]
```diff
@@ -104,73 +137,139 @@ func New(conf Config, eb *appevent.Broadcaster) *Network {
...
 		}
 	}
 
+	if n.conf.NetworkConfigs.STCPR != nil {
+		if n.clients.StcprC != nil && n.conf.NetworkConfigs.STCPR.LocalAddr != "" {
+			if err := n.clients.StcprC.Serve(); err != nil {
+				return fmt.Errorf("failed to initiate 'stcpr': %w", err)
+			}
+		} else {
+			fmt.Println("No config found for stcpr")
```
Comment (Darkren) [inline pkg/snet/network.go]: Shouldn't we use logger here?

File: pkg/snet/stcph/client.go [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+	}
+
+	c.once.Do(func() {
+		close(c.done)
+
+		c.mx.Lock()
+		defer c.mx.Unlock()
+
+		// TODO: log error
+		_ = c.addressResolver.Close() // nolint:errcheck
```
Comment (Darkren) [inline pkg/snet/stcph/client.go]: should we log?

File: pkg/snet/stcph/conn.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+	hs        Handshake
+	freePort  func()
+	encrypt   bool
+	initiator bool
+}
+
+func newConn(c connConfig) (*Conn, error) {
+	lAddr, rAddr, err := c.hs(c.conn, c.deadline)
+	if err != nil {
+		_ = c.conn.Close() //nolint:errcheck
```
Comment (Darkren) [inline pkg/snet/stcph/conn.go]: should we log it?

File: pkg/snet/stcph/handshake.go [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+		if err = conn.SetDeadline(deadline); err != nil {
+			return
+		}
+
+		if lAddr, rAddr, err = origin(conn, deadline); err != nil {
+			err = HandshakeError(err.Error())
+		}
+
+		// reset deadline
+		_ = conn.SetDeadline(time.Time{}) //nolint:errcheck
```
Comment (Darkren) [inline pkg/snet/stcph/handshake.go]: should we log it?

File: pkg/snet/stcpr/client.go [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+	}
+
+	c.once.Do(func() {
+		close(c.done)
+
+		c.mx.Lock()
+		defer c.mx.Unlock()
+
+		if c.lTCP != nil {
+			_ = c.lTCP.Close() //nolint:errcheck
```
Comment (Darkren) [inline pkg/snet/stcpr/client.go]: should we log it? and the one a couples of lines below

File: pkg/snet/stcpr/conn.go [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+	hs        Handshake
+	freePort  func()
+	encrypt   bool
+	initiator bool
+}
+
+func newConn(c connConfig) (*Conn, error) {
+	lAddr, rAddr, err := c.hs(c.conn, c.deadline)
+	if err != nil {
+		_ = c.conn.Close() //nolint:errcheck
```
Comment (Darkren) [inline pkg/snet/stcpr/conn.go]: should we log?

File: pkg/snet/stcpr/handshake.go [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+		if err = conn.SetDeadline(deadline); err != nil {
+			return
+		}
+
+		if lAddr, rAddr, err = origin(conn, deadline); err != nil {
+			err = HandshakeError(err.Error())
+		}
+
+		// reset deadline
+		_ = conn.SetDeadline(time.Time{}) //nolint:errcheck
```
Comment (Darkren) [inline pkg/snet/stcpr/handshake.go]: should we log?

File: pkg/snet/stcpr/handshake.go [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+// Verify verifies the signature field within Frame2.
+func (f2 Frame2) Verify(nonce [HandshakeNonceSize]byte) error {
+	if f2.Nonce != nonce {
+		return errors.New("unexpected nonce")
+	}
+
+	sig := f2.Sig
+	f2.Sig = cipher.Sig{}
+
+	//cipher2.PubKeyFromSig(cipher2.Sig(sig))
```
Comment (Darkren) [inline pkg/snet/stcpr/handshake.go]: should we remove this line?

File: pkg/snet/stcpr/handshake.go [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+
+	var b bytes.Buffer
+	if err := json.NewEncoder(&b).Encode(f2); err != nil {
+		return err
+	}
+
+	hash := cipher.SumSHA256(b.Bytes())
+	rPK, err := cipher2.PubKeyFromSig(cipher2.Sig(sig), cipher2.SHA256(hash))
+
+	fmt.Println("VERIFY! len(b.Bytes)", len(b.Bytes()), cipher2.SHA256(hash).Hex(),
```
Comment (Darkren) [inline pkg/snet/stcpr/handshake.go]: should this one be removed? or at least moved to log with the `debug` level

**Review by Darkren — CHANGES_REQUESTED** [review]
Overall, huge job. Comments are mostly stylish

File: README.md (hunk 1 above)
Comment (nkryuchkov) [reply README.md]: As `stcp` is not removed, these changes need to be reverted. Well spotted! Thanks

File: internal/vpn/env.go (hunk 3 above)
Comment (nkryuchkov) [reply internal/vpn/env.go]: Sure, I was going to name it as you suggested but something went wrong.

File: pkg/snet/arclient/client.go (hunk 4 above)
Comment (nkryuchkov) [reply pkg/snet/arclient/client.go]: I copied the code from another client.

File: pkg/snet/network.go (hunk 6 above)
Comment (nkryuchkov) [reply pkg/snet/network.go]: I agree. I copied this from `stcp` and didn't change anything. The same applies to your comments below where there's no reply from me.

File: pkg/snet/stcph/client.go (hunk 7 above)
Comment (nkryuchkov) [reply pkg/snet/stcph/client.go]: Yes, this should be logged.

**Comment by nkryuchkov** [comment]
@Darkren Thanks for reviewing. Your requests are resolved.

**Pull request merged**